File: 4/4.cpp
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
	if (nums1.size() > nums2.size()) {
		swap(nums1, nums2);
	}

	int m = nums1.size(), n = nums2.size();

	int imin = 0, imax = m;
	while (imin <= imax) {
		int i = imin + (imax - imin) / 2;
		int j = (m + n + 1) / 2 - i;
		printf("i %d\n", i);
		printf("j %d\n", j);

		if (i > 0 && nums1[i - 1] > nums2[j]) {
			imax = i - 1;
		}
		else if (i < m && nums2[j - 1] > nums1[i]) {
			imin = i + 1;
		}
        else{
			if ((m + n) % 2) {
				if (i == 0)
					return nums2[j - 1];
				if (j == 0)
					return nums1[i - 1];
				return nums1[i - 1] > nums2[j - 1] ? nums1[i - 1] : nums2[j - 1];
			}
			else {
				int l = INT_MIN, r = INT_MAX;
                if(m == 0) return (nums2[j]+nums2[j-1])/2.0;
                
                if(i > 0) l = max(l, nums1[i-1]);
                if(j > 0) l = max(l, nums2[j-1]);
                if(i < m) r = min(r, nums1[i]);
                if(j < n) r = min(r, nums2[j]);
                
                return (l+r)/2.0;
			}
		}
	}
	return 0;
}
};

File: 4/4.rs
impl Solution {
    pub fn find_median_sorted_arrays(nums1: Vec<i32>, nums2: Vec<i32>) -> f64 {
        let short;
        let long;
        if nums1.len() <= nums2.len() {
            short = nums1;
            long = nums2;
        } else {
            short = nums2;
            long = nums1;
        }

        let (mut low, mut high) = (0, short.len());
        let totalLen = short.len() + long.len();

        while low <= high {
            let sx = low + (high - low) / 2;
            let lx = (short.len() + long.len()) / 2 - sx;

            let sleft = {
                if sx > 0 {
                    short[sx - 1]
                } else {
                    std::i32::MIN
                }
            };
            let sright = {
                if sx < short.len() {
                    short[sx]
                } else {
                    std::i32::MAX
                }
            };
            let lleft = {
                if lx > 0 {
                    long[lx - 1]
                } else {
                    std::i32::MIN
                }
            };
            let lright = {
                if lx < long.len() {
                    long[lx]
                } else {
                    std::i32::MAX
                }
            };

            if sleft <= lright && lleft <= sright {
                if totalLen % 2 == 0 {
                    let left = {
                        if sleft >= lleft {
                            sleft
                        } else {
                            lleft
                        }
                    };
                    let right = {
                        if sright <= lright {
                            sright
                        } else {
                            lright
                        }
                    };
                    return (left as f64 + right as f64) / 2.0;
                } else {
                    if sright <= lright {
                        return sright as f64;
                    } else {
                        return lright as f64;
                    }
                }
            }

            if sleft > lright {
                high = sx;
                continue;
            }

            if lleft > sright {
                low = sx + 1;
                continue;
            }
        }
        panic!("Wrong arrays");
    }
}